core/base_abstractions/sensor.py

```diff
@@ -158,23 +166,34 @@ def _get_observation_space(self) -> gym.spaces.Tuple:
         only if the expert failed to generate a true expert action. The
         value `num actions in task` should be in `config["nactions"]`
         """
-        return gym.spaces.Tuple(
-            (gym.spaces.Discrete(self.nactions), gym.spaces.Discrete(2))
-        )
+        return su.flatten_space(self.unflattened_observation_space)
 
     def get_observation(
         self, env: EnvType, task: SubTaskType, *args: Any, **kwargs: Any
     ) -> Any:
         # If the task is completed, we needn't (perhaps can't) find the expert
         # action from the (current) terminal state.
         if task.is_done():
-            return np.array([-1, False], dtype=np.int64)
+            return np.array([self.action_space.sample(), False])
+
         action, expert_was_successful = task.query_expert(**self.expert_args)
-        assert isinstance(action, int), (
-            "In expert action sensor, `task.query_expert()` "
-            "did not return an integer action."
+
+        if isinstance(action, int):
+            assert isinstance(self.action_space, gym.spaces.Discrete)
+            unflattened_action = action
+        else:
+            # Assume we receive a gym-flattened numpy action
+            unflattened_action = gyms.unflatten(self.action_space, action)
+
+        unflattened_torch = su.torch_point(
+            self.unflattened_observation_space,
+            (unflattened_action, expert_was_successful),
+        )
+
+        flattened_torch = su.flatten(
+            self.unflattened_observation_space, unflattened_torch
         )
-        return np.array([action, expert_was_successful], dtype=np.int64)
+        return flattened_torch.cpu().numpy()
```

Comment:
I'm going to make a slightly more radical suggestion about expert actions. I think we should change the current `Tuple` output space to instead be something like:
```
gym.spaces.Dict({
    "action": ...,
    "mask": ...
})
```
I think this is more intuitive and what we would have done from the start if our original design had allowed for dictionary output spaces.

Reply:
We can do this after merging your PR though assuming this works well with the current design.

Reply:
Yes, that's probably a better option. I just tried to reuse code in engine where possible, but it's going to be cleaner. The limitation that I see is that if the action contains Tuples, it won't be managed by the current Memory (which is used as storage for rollouts).

Reply:
Sounds good, let's keep this for after this PR since it's more or less its own confined change.